### DISCUSSION TITLE
Using `next/link` with `semantic-ui-react/Card` and other non-`a` components

### DISCUSSION THREAD

**Discussion by AlbinoGeek** [vercel/next.js#25606] [Help] [answered]
**I hope I missed something really obvious / easy somewhere in the docs and this was simply my mistake.**

**TLDR:** See Title.

## Summary

1. Wrapping `<Link href={string}>` around a `<Card link>` does not behave as expected (might be documented).
2. Wrapping `<Link passHref href={string}>` around a `<Card link>` does not behave as expected.

**Observation:** Semantic-UI `<Card link>` renders to an `<a>`

## Expectation

A clickable link-styled Card that asks the router to politely transition the user to the requested page.

## Reality

1. A clickable link-styled blob of misrendered garble which resolves in-app 404 or full-page navigation.
    HTML Output is `<a href="..."><a class="ui card ...">GARBLED</a></a>`

2. A clickable link-styled correctly rendered Semantic-UI Card which resolves in-app 404 or full-page navigation.
    HTML Output is `<a href="..." class="ui card ...">... [renders correctly] ...</a>`

1 and 2: The `Card` outputs as an `<a>` tag with an href but no `onClick`, and as such, causes full-page navigation.

---

After immense trial and error, the following code sample "works" (full SSR, working PWA, client transitions, everything)...  But it's fairly evident this approach is far-from-ideal and error-prone.

### `/package.json`

```json
  "dependencies": {
    "next": "^10.2.2",
    "next-pwa": "^5.2.21",
    "react": "^17.0.6",
    "react-dom": "^17.0.3",
    "react-is": "^17.0.2",
    "semantic-ui-react": "^2.0.3",
  },
  "devDependencies": {
    "@babel/core": "^7.11.1",
    "@babel/eslint-parser": "^7.14.4",
    "@types/node": "^15.0.0",
    "@types/react": "^17.0.6",
    "@types/react-dom": "^17.0.3",
    "@typescript-eslint/eslint-plugin": "^4.25.0",
    "@typescript-eslint/parser": "^4.25.0",
    "eslint": "^7.25.0",
    "eslint-plugin-jsx-a11y": "^6.4.1",
    "eslint-plugin-react": "^7.23.2",
    "eslint-plugin-react-hooks": "^4.2.0",
    "fomantic-ui": "^2.8.7",
    "inquirer": "^7.3.3",
    "typescript": "^4.3.2",
    "webpack": "^5.37.1"
  },
```

### `/components/Product.tsx`

```tsx
// eslint changes all my `next/x` imports into `next/dist/client/x` -_-
import { useRouter } from 'next/dist/client/router'
import { Card, Icon, Placeholder } from 'semantic-ui-react'
import Category from '../types/Category'
import Product from '../types/Product'

type Props = {
  category?: Category
  product: Product
}

const ProductCard = ({ category, product }: Props): JSX.Element => {
  const href = `/product/${product.ID}`
  const router = useRouter()
  
  // TODO: jsdoc shows `MouseEvent<HTMLAnchorElement, MouseEvent>` but ESLint says that's invalid?
  const navigate = (e: { preventDefault: () => void }) => {
    e.preventDefault()
    router.push(href)
  }
  
  // Would use `next/link` here, but doesn't appear to work as expected.
  return (
    <Card link href={href} onClick={navigate}>
      <Placeholder>
        <Placeholder.Image rectangular />
      </Placeholder>
    </Card>
  )
}
```

### `/pages/product/[id].tsx`

**Pretty much a carbon-copy of the example found in the repository at the URL as follows:**
https://github.com/vercel/next.js/blob/canary/examples/with-typescript/pages/users/%5Bid%5D.tsx

```tsx
import { GetStaticPaths, GetStaticProps } from 'next'
import ProductCard from '../../components/Product'
import { GetAllProducts } from '../../store'
import Product from '../../types/Product'
import Error400 from '../400'
import Error404 from '../404'

const StaticPropsDetail = (props: { item?: Product, error?: string }): JSX.Element => {
  if (error) return <Error400 message={error} />
  if (item === undefined) return <Error404 />
  return <ProductCard product={item} />
}

export default StaticPropsDetail

export const getStaticPaths: GetStaticPaths = async () => {
  const products = await GetAllProducts()
  const paths = products.map((product) => ({
    params: { id: product.ID.toString() },
  }))

  return { paths, fallback: false }
}

export const getStaticProps: GetStaticProps = async ({ params }) => {
  try {
    const id = params?.id
    const products = await GetAllProducts()
    const item = products.find((product) => product.ID === Number(id))
    return { props: { item } }
  } catch (err) {
    return { props: { errors: err.message } }
  }
}
```

---

I will gladly add more information about my application if these issues aren't easily reproducible as shown.

---

**Edit**: Simplified code example after futzing around with parameters/href/as

**Comment by wwwdepot** [comment, marked as answer]
Wrap your `ProductCard` in `React.forwardRef` https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component

**Reply by AlbinoGeek** [reply]
> Wrap your `ProductCard` in `React.forwardRef` https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component

I mean... it only says I have to do that when wrapping the link around a functional component.

My `ProductCard` is responsible for its own Link.

The child of the `ProductCard` is `Card` which is not a functional component (it's a class component.)

---

Wrapping (by way of `forwardRef` and `Ref` with `innerRef`) [such as I have to do with `framer-motion`] did not fix this.

```tsx
import { motion } from 'framer-motion'
import { useRouter } from 'next/dist/client/router'
import React, { ForwardedRef, forwardRef } from 'react'
import { Card, Placeholder, Ref } from 'semantic-ui-react'
import type { Category } from '../types/Category'
import type { Product } from '../types/Product'

type Props = {
  readonly category?: Category
  readonly product: Product
}

const ProductCard = forwardRef((
  { category, product }: Props,
  ref: ForwardedRef<HTMLElement>
  ): JSX.Element => {
  const href = `/product/${product.ID}`
  const router = useRouter()
  
  // still required, even with forwardRef / ForwardedRef
  const navigate = (e: { preventDefault: () => void }) => {
    e.preventDefault()
    router.push(href)
  }
  
  // Wrapping the Ref in a Link does not fix the Link
  return <Ref innerRef={ref}>
    <Card link href={href} onClick={navigate}>
      <Placeholder fluid>
        <Placeholder.Image rectangular />
      </Placeholder>
    </Card>
  </Ref>
})

// NB: I hate TypeScript / ESLint sometimes...
ProductCard.displayName = 'ProductCard'

// Wrapping this whole ProductCard in a Link does not fix the Link
export default motion(ProductCard)

```

**Reply by AlbinoGeek** [reply]
This eventually led to a solution, the other being to stop using semantic, since the Drawer (well, Pushable specifically) spews lots of errors in React Strict.
